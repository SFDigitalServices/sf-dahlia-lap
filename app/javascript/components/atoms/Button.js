import React from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import { toPx } from 'utils/cssUtils'

const buttonSizeStyles = (clearVerticalPadding, minWidth) => {
  const widthStyle = minWidth
    ? {
        minWidth: toPx(minWidth)
      }
    : {}

  const heightStyle = clearVerticalPadding
    ? {
        paddingTop: 0,
        paddingBottom: 0
      }
    : {}

  return {
    ...heightStyle,
    ...widthStyle
  }
}

const buttonChildrenStyles = () => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%'
})

const textStyles = (hasLeftIcon, hasRightIcon) => ({
  marginTop: '1rem',
  marginBottom: '1rem',
  marginRight: hasRightIcon && '1rem',
  marginLeft: hasLeftIcon && '1rem',
  flexGrow: 1
})

const iconWrapperStyles = () => ({
  flexShrink: 0,

  // we need to override font-size because of a css button selector that sets
  // the font size for all children. This is a problem because it adds 3px
  // to any icon wrapper.
  fontSize: 0
})

const wrapWithStyle = (node, style) => node && <span style={style}>{node}</span>

const Button = ({
  classes = null,
  iconLeft = null,
  iconRight = null,
  noBottomMargin,
  paddingHorizontal = 'normal',
  small = false,
  tertiary = false,
  text = null,
  tiny = false,
  type = 'button',
  minWidthPx = null,
  style = {},
  ...rest
}) => {
  const btnClassNames = classNames(classes, {
    'tight-padding': paddingHorizontal === 'tight',
    'extra-padding': paddingHorizontal === 'extra',
    small: small,
    tertiary: tertiary,
    tiny: tiny,
    'margin-bottom-none': noBottomMargin
  })

  const hasText = !!text

  return (
    <button
      className={btnClassNames}
      type={type}
      style={{ ...buttonSizeStyles(hasText, minWidthPx), ...style }}
      {...rest}
    >
      <div style={buttonChildrenStyles()}>
        {wrapWithStyle(iconLeft, iconWrapperStyles())}
        {wrapWithStyle(text, textStyles(!!iconLeft, !!iconRight))}
        {wrapWithStyle(iconRight, iconWrapperStyles())}
      </div>
    </button>
  )
}

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit']),
  noBottomMargin: PropTypes.bool,
  classes: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  iconLeft: PropTypes.node,
  iconRight: PropTypes.node,
  minWidthPx: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingHorizontal: PropTypes.oneOf(['normal', 'tight', 'extra']),
  small: PropTypes.bool,
  style: PropTypes.object,
  tertiary: PropTypes.bool,
  text: PropTypes.node,
  tiny: PropTypes.bool
}

export default Button
