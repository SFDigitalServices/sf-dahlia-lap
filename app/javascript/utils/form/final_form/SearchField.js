import React, { useRef } from 'react'
import PropTypes from 'prop-types'

import { Field } from 'react-final-form'
import classNames from 'classnames'
import { Label, FieldError, HelpText } from './Field'

const SearchField = ({
  blockNote,
  disabled = false,
  fieldName,
  format,
  formatOnBlur = false,
  helpText,
  id,
  isDirty = false,
  label,
  maxLength,
  onClearClick,
  placeholder,
  validation
}) => {
  const inputElement = useRef(null)

  return (
    <Field
      name={fieldName}
      validate={validation}
      format={format}
      formatOnBlur={formatOnBlur && isDirty}
    >
      {({ input, meta }) => {
        const hasValue = !!input.value
        return (
          <>
            <div
              className={classNames('search', {
                'form-group': label,
                error: meta.error && meta.touched
              })}
            >
              <Label
                label={label}
                id={id || `form-${fieldName}`}
                fieldName={fieldName}
                blockNote={blockNote}
              />
              <input
                {...input}
                id={id || `form-${fieldName}`}
                className={classNames({ 'has-value': hasValue, error: meta.error && meta.touched })}
                type='text'
                maxLength={maxLength}
                placeholder={placeholder}
                disabled={disabled}
                ref={inputElement}
              />
              {hasValue && (
                <button
                  className='ui-medium i-aluminum button-link search-icon'
                  onClick={() => {
                    onClearClick()
                    inputElement.current.focus()
                  }}
                  type='button'
                >
                  <svg>
                    <use xlinkHref={`#i-close-round`} />
                  </svg>
                </button>
              )}
              <FieldError meta={meta} />
              {helpText && <HelpText note={helpText} describeId={`describe-${id || fieldName}`} />}
            </div>
          </>
        )
      }}
    </Field>
  )
}

SearchField.propTypes = {
  blockNote: PropTypes.string,
  disabled: PropTypes.bool,
  fieldName: PropTypes.string.isRequired,
  format: PropTypes.func,
  formatOnBlur: PropTypes.bool,
  helpText: PropTypes.string,
  id: PropTypes.string,
  isDirty: PropTypes.bool,
  label: PropTypes.string,
  maxLength: PropTypes.number,
  onClearClick: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  validation: PropTypes.func
}
export default SearchField
