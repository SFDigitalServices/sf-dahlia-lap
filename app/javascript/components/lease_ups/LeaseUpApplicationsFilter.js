import React from 'react'
import { Form } from 'react-final-form'
import { InputField, SelectField } from '~/utils/form/final_form/Field'
import formUtils from '~/utils/formUtils'
import LEASE_UP_STATUS_OPTIONS, { LEASE_UP_ACCESSIBILITY_OPTIONS } from '~/utils/statusUtils'
import Loading from '~/components/molecules/Loading'
import { clone } from 'lodash'

const statusFilterOptions = [
  ...[{value: null, label: 'Any Status'}, {value: 'No Status', label: 'No Status'}],
  ...clone(LEASE_UP_STATUS_OPTIONS)
]

const accessibilityFilterOptions = [
  {value: null, label: 'Any Accessibility Request'},
  ...clone(LEASE_UP_ACCESSIBILITY_OPTIONS)
]

const StatusOptions = formUtils.toOptions(statusFilterOptions)

const LeaseUpApplicationsFilter = ({ onSubmit, preferences = [], loading = false }) => {
  const preferenceOptions = formUtils.toOptions([[null, 'Any Preference'], ...preferences, ['general', 'General']])

  return (
    <Loading isLoading={loading}>
      <Form
        onSubmit={onSubmit}
        render={({ handleSubmit }) => (
          <form onSubmit={handleSubmit} noValidate>
            <div className='filter-row'>
              <div className='filter-group'>
                <div className='filter-group_item'>
                  <SelectField fieldName='preference' options={preferenceOptions} placeholder='Preference' />
                </div>
                <div className='filter-group_item'>
                  <InputField fieldName='application_number' placeholder='Application Number' />
                </div>
                <div className='filter-group_item'>
                  <InputField fieldName='first_name' placeholder='First Name' />
                </div>
                <div className='filter-group_item'>
                  <InputField fieldName='last_name' placeholder='Last Name' />
                </div>
                <div className='filter-group_item'>
                  <SelectField fieldName='status' options={StatusOptions} placeholder='Status' />
                </div>
                <div className='filter-group_item'>
                  <SelectField fieldName='accessibility' options={accessibilityFilterOptions} placeholder='Accessibility Requests' />
                </div>
                <div className='filter-group_action'>
                  <button className='small'>Filter</button>
                </div>
              </div>
            </div>
          </form>
        )} />
    </Loading>
  )
}

export default LeaseUpApplicationsFilter
