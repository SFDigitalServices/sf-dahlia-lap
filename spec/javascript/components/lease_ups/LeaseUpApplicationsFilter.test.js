import React from 'react'
import { mount, shallow } from 'enzyme'

import LeaseUpApplicationsFilter from '~/components/lease_ups/LeaseUpApplicationsFilter'
import SearchField from '~/utils/form/final_form/SearchField'
import Button from '~/components/atoms/Button'

const mockSubmit = jest.fn()

const getShallowWrapper = () =>
  shallow(
    <LeaseUpApplicationsFilter
      onSubmit={mockSubmit}
      preferences={['pref option 1', 'pref option 2']}
    />
  )

const getWrapper = () =>
  mount(
    <LeaseUpApplicationsFilter
      onSubmit={mockSubmit}
      preferences={['pref option 1', 'pref option 2']}
    />
  )

describe('LeaseUpApplicationsFilter', () => {
  test('it matches the expected snapshot with default preferences', () => {
    const wrapper = getShallowWrapper()
    expect(wrapper.find('ReactFinalForm').dive()).toMatchSnapshot()
  })

  describe('search field', () => {
    test('it does not show the clear button when there is no text in the field', () => {
      const wrapper = getWrapper()
      expect(wrapper.exists('.search-icon')).toBe(false)
    })
    test('it renders a clear input icon when there is text in the search input', () => {
      const wrapper = getWrapper()
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(wrapper.exists('.search-icon')).toBe(true)
    })
    test('the clear input button clears the input', () => {
      const wrapper = getWrapper()
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(wrapper.find(SearchField).find('input').props().value).toEqual('my search query')

      wrapper.find('.search-icon').simulate('click')
      expect(wrapper.find(SearchField).find('input').props().value).toEqual('')
    })
    test('the search field is disabled unless the field has been modified', () => {
      const wrapper = getWrapper()
      const searchButtonIsDisabled = (wrapper) =>
        wrapper.find(Button).find({ text: 'Search' }).find('button').props().disabled
      // search button is disabled on first render
      expect(searchButtonIsDisabled(wrapper)).toBe(true)

      // search button should be enabled after we modify the field
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(searchButtonIsDisabled(wrapper)).toBe(false)

      // search button should be disabled after we submit
      console.log('clicking submit')
      wrapper.find(Button).find({ text: 'Search' }).find('button').simulate('submit')
      expect(searchButtonIsDisabled(wrapper)).toBe(true)

      // search button should be re-enabled if we modify after submit
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my other search query' } })
      wrapper.update()
      expect(searchButtonIsDisabled(wrapper)).toBe(false)
    })
  })
})
