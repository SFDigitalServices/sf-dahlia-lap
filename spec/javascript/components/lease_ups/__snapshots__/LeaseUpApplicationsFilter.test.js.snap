// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseUpApplicationsFilter it matches the expected snapshot with default preferences 1`] = `
<ContextProvider
  value={
    Object {
      "batch": [Function],
      "blur": [Function],
      "change": [Function],
      "destroyOnUnregister": false,
      "focus": [Function],
      "getFieldState": [Function],
      "getRegisteredFields": [Function],
      "getState": [Function],
      "initialize": [Function],
      "isValidationPaused": [Function],
      "mutators": Object {},
      "pauseValidation": [Function],
      "registerField": [Function],
      "reset": [Function],
      "resetFieldState": [Function],
      "restart": [Function],
      "resumeValidation": [Function],
      "setConfig": [Function],
      "submit": [Function],
      "subscribe": [Function],
    }
  }
>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="filter-row"
    >
      <div
        className="filter-group"
      >
        <div
          className="filter-group_item"
        >
          <SearchField
            fieldName="search"
            id="test-search"
            onClearClick={[Function]}
            placeholder="Application, First Name, Last Name..."
          />
        </div>
        <div
          className="filter-group_action"
        >
          <Button
            className="small primary"
            disabled={true}
            text="Search"
            type="submit"
          />
        </div>
        <div
          className="filter-group_item"
        >
          <SelectField
            fieldName="preference"
            options={
              Array [
                Object {
                  "label": "Any Preference",
                  "value": "",
                },
                Object {
                  "label": "pref option 1",
                  "value": "pref option 1",
                },
                Object {
                  "label": "pref option 2",
                  "value": "pref option 2",
                },
                Object {
                  "label": "General",
                  "value": "general",
                },
              ]
            }
            placeholder="Preference"
          />
        </div>
        <div
          className="filter-group_item"
        >
          <SelectField
            fieldName="total_household_size"
            options={
              Array [
                Object {
                  "label": "Any HH Members",
                  "value": "",
                },
                Object {
                  "label": "1",
                  "value": "1",
                },
                Object {
                  "label": "2",
                  "value": "2",
                },
                Object {
                  "label": "3",
                  "value": "3",
                },
                Object {
                  "label": "4",
                  "value": "4",
                },
                Object {
                  "label": "5+",
                  "value": "5+",
                },
              ]
            }
          />
        </div>
        <div
          className="filter-group_item"
        >
          <SelectField
            fieldName="status"
            options={
              Array [
                Object {
                  "label": "Any Status",
                  "value": "",
                },
                Object {
                  "label": "No Status",
                  "value": "No Status",
                },
                Object {
                  "commentRequired": true,
                  "label": "Processing",
                  "statusClassName": "is-processing",
                  "value": "Processing",
                },
                Object {
                  "label": "Withdrawn",
                  "statusClassName": "is-withdrawn",
                  "value": "Withdrawn",
                },
                Object {
                  "commentRequired": true,
                  "label": "Appealed",
                  "statusClassName": "is-appealed",
                  "value": "Appealed",
                },
                Object {
                  "commentRequired": true,
                  "label": "Waitlisted",
                  "statusClassName": "is-waitlisted",
                  "value": "Waitlisted",
                },
                Object {
                  "commentRequired": true,
                  "label": "Disqualified",
                  "statusClassName": "is-disqualified",
                  "value": "Disqualified",
                },
                Object {
                  "commentRequired": true,
                  "label": "Approved",
                  "statusClassName": "is-approved",
                  "value": "Approved",
                },
                Object {
                  "label": "Lease Signed",
                  "statusClassName": "is-leased",
                  "value": "Lease Signed",
                },
              ]
            }
            placeholder="Status"
          />
        </div>
        <div
          className="filter-group_item"
        >
          <SelectField
            fieldName="accessibility"
            options={
              Array [
                Object {
                  "label": "Any Accessibility Request",
                  "value": "",
                },
                Object {
                  "label": "Mobility",
                  "value": "Mobility impairments",
                },
                Object {
                  "label": "Vision/Hearing",
                  "value": "Vision impairments, Hearing impairments",
                },
              ]
            }
            placeholder="Accessibility Requests"
          />
        </div>
        <div
          className="filter-group_action"
        >
          <button
            className="small"
          >
            Filter
          </button>
        </div>
      </div>
    </div>
  </form>
</ContextProvider>
`;
